File: sdk/typescript/tests/tools.test.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 */

import { describe, expect, it } from "vitest";

import { InvalidArgumentsError, Plinth, ToolNotFoundError } from "../src/index.js";
import { MockServer, parseQuery } from "./_helpers.js";

function makeClient(server: MockServer): Plinth {
  return new Plinth({
    workspaceUrl: "http://workspace.test",
    gatewayUrl: "http://gateway.test",
    apiKey: "test-token",
    fetch: server.fetch as unknown as typeof fetch,
  });
}

describe("ToolsClient", () => {
  it("invokes a tool and returns the gateway response", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/invoke$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      expect(body.tool_id).toBe("web.fetch");
      expect(body.arguments).toEqual({ url: "https://example.com" });
      return {
        body: {
          tool_id: "web.fetch",
          arguments: body.arguments,
          result: { content: "<!doctype html>", status: 200, content_type: "text/html" },
          cached: false,
          duration_ms: 42,
          audit_id: "evt_xyz",
          cost_estimate_usd: 0.0,
        },
      };
    });

    const client = makeClient(server);
    const res = await client.tools.invoke("web.fetch", { url: "https://example.com" });
    expect(res.tool_id).toBe("web.fetch");
    expect(res.cached).toBe(false);
    expect((res.result as { status: number }).status).toBe(200);
  });

  it("forwards workspace_id and agent_id for audit attribution", async () => {
    const server = new MockServer();
    let captured: Record<string, unknown> = {};
    server.on("POST", /\/v1\/invoke$/, (req) => {
      captured = JSON.parse(req.body ?? "{}");
      return {
        body: {
          tool_id: "web.fetch",
          arguments: {},
          result: null,
          cached: false,
          duration_ms: 1,
          audit_id: "evt_a",
          cost_estimate_usd: 0,
        },
      };
    });

    const client = makeClient(server);
    await client.tools.invoke(
      "web.fetch",
      { url: "u" },
      { workspaceId: "ws_1", agentId: "agent_1", cache: false, idempotencyKey: "k1" },
    );
    expect(captured).toMatchObject({
      tool_id: "web.fetch",
      workspace_id: "ws_1",
      agent_id: "agent_1",
      cache: false,
      idempotency_key: "k1",
    });
  });

  it("dryRun POSTs to /v1/invoke/dry-run", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/invoke\/dry-run$/, () => ({
      body: {
        tool_id: "web.fetch",
        arguments: {},
        would_invoke: false,
        cached_result: { content: "" },
        estimated_cost_usd: 0,
        estimated_duration_ms: 0,
      },
    }));

    const client = makeClient(server);
    const res = await client.tools.dryRun("web.fetch", { url: "x" });
    expect(res.would_invoke).toBe(false);
    expect(res.cached_result).toEqual({ content: "" });
  });

  it("register POSTs the tool registration body", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tools\/register$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      return {
        status: 201,
        body: {
          ...body,
          created_at: "2026-01-01T00:00:00Z",
          updated_at: "2026-01-01T00:00:00Z",
        },
      };
    });

    const client = makeClient(server);
    const tool = await client.tools.register({
      tool_id: "test.echo",
      name: "echo",
      description: "echoes input",
      transport: "http",
      endpoint: "http://localhost:9999",
      input_schema: {},
      output_schema: {},
    });
    expect(tool.tool_id).toBe("test.echo");
    expect(tool.created_at).toBe("2026-01-01T00:00:00Z");
  });

  it("list returns the tools array", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/tools$/, {
      tools: [
        {
          tool_id: "web.fetch",
          name: "fetch",
          description: "",
          transport: "http",
          endpoint: "x",
          input_schema: {},
          output_schema: {},
          created_at: "2026-01-01T00:00:00Z",
          updated_at: "2026-01-01T00:00:00Z",
        },
      ],
    });

    const client = makeClient(server);
    const tools = await client.tools.list();
    expect(tools).toHaveLength(1);
    expect(tools[0]!.tool_id).toBe("web.fetch");
  });

  it("audit forwards query parameters", async () => {
    const server = new MockServer();
    server.on("GET", /\/v1\/audit/, (req) => {
      const q = parseQuery(req.url);
      expect(q.workspace_id).toBe("ws_1");
      expect(q.since).toBe("1h");
      return { body: { events: [] } };
    });

    const client = makeClient(server);
    const events = await client.tools.audit({ workspaceId: "ws_1", since: "1h" });
    expect(events).toEqual([]);
  });

  it("maps TOOL_NOT_FOUND to typed error", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/invoke$/, () => ({
      status: 404,
      body: { error: { code: "TOOL_NOT_FOUND", message: "no such tool" } },
    }));
    const client = makeClient(server);
    await expect(client.tools.invoke("missing", {})).rejects.toBeInstanceOf(ToolNotFoundError);
  });

  it("maps INVALID_ARGUMENTS to typed error", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/invoke$/, () => ({
      status: 400,
      body: { error: { code: "INVALID_ARGUMENTS", message: "missing url", details: { field: "url" } } },
    }));
    const client = makeClient(server);
    const err = await client.tools
      .invoke("web.fetch", {})
      .catch((e: unknown) => e as InvalidArgumentsError);
    expect(err).toBeInstanceOf(InvalidArgumentsError);
    expect((err as InvalidArgumentsError).details).toEqual({ field: "url" });
  });
});

File: sdk/go/plinth_test/tools_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth_test

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/plinth/sdk-go/plinth"
)

// TestToolsInvokeHappyPath covers the basic invoke flow: arguments
// flow through, the response decodes into InvokeResponse, and the
// returned cached/duration_ms fields are surfaced.
func TestToolsInvokeHappyPath(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("POST", "/v1/invoke", 200, map[string]any{
		"tool_id":   "web.fetch",
		"arguments": map[string]any{"url": "mock://test"},
		"result":    map[string]any{"content": "hello", "status": 200},
		"cached":    false,
		"duration_ms": 12,
		"audit_id":  "evt_abc",
		"cost_estimate_usd": 0.0,
	})
	c := newTestClient(t, ms)

	resp, err := c.Tools.Invoke(context.Background(), "web.fetch",
		map[string]any{"url": "mock://test"},
		plinth.InvokeOpts{},
	)
	if err != nil {
		t.Fatalf("Invoke: %v", err)
	}
	if resp.ToolID != "web.fetch" {
		t.Errorf("resp.ToolID = %q, want web.fetch", resp.ToolID)
	}
	if resp.AuditID != "evt_abc" {
		t.Errorf("resp.AuditID = %q, want evt_abc", resp.AuditID)
	}
	if resp.Cached {
		t.Errorf("resp.Cached = true, want false")
	}

	// Verify the request body shape.
	last := ms.LastRequest(t)
	var body map[string]any
	last.jsonBody(t, &body)
	if body["tool_id"] != "web.fetch" {
		t.Errorf("body.tool_id = %v, want web.fetch", body["tool_id"])
	}
}

// TestToolsInvokeCachedFlag verifies the response signals a cache hit.
func TestToolsInvokeCachedFlag(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("POST", "/v1/invoke", 200, map[string]any{
		"tool_id":   "web.fetch",
		"arguments": map[string]any{"url": "mock://test"},
		"result":    map[string]any{"content": "cached"},
		"cached":    true,
		"duration_ms": 0,
		"audit_id":  "evt_def",
		"cost_estimate_usd": 0.0,
	})
	c := newTestClient(t, ms)
	resp, err := c.Tools.Invoke(context.Background(), "web.fetch",
		map[string]any{"url": "mock://test"},
		plinth.InvokeOpts{},
	)
	if err != nil {
		t.Fatalf("Invoke: %v", err)
	}
	if !resp.Cached {
		t.Errorf("resp.Cached = false, want true")
	}
}

// TestToolsInvokeNotFound covers the 404 path on POST /v1/invoke.
// errors.Is should match ErrToolNotFound.
func TestToolsInvokeNotFound(t *testing.T) {
	ms := NewMockServer(t)
	ms.Error("POST", "/v1/invoke", 404, "TOOL_NOT_FOUND", "no such tool")
	c := newTestClient(t, ms)

	_, err := c.Tools.Invoke(context.Background(), "no.such",
		map[string]any{}, plinth.InvokeOpts{})
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !errors.Is(err, plinth.ErrToolNotFound) {
		t.Errorf("err = %v, want ErrToolNotFound", err)
	}
}

// TestToolsInvokeCacheDisabledIsForwarded checks that InvokeOpts.Cache
// = BoolPtr(false) ends up in the request body so the gateway can act
// on it.
func TestToolsInvokeCacheDisabledIsForwarded(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("POST", "/v1/invoke", 200, map[string]any{
		"tool_id": "web.fetch", "arguments": map[string]any{}, "result": "ok",
		"cached": false, "duration_ms": 1, "audit_id": "x", "cost_estimate_usd": 0.0,
	})
	c := newTestClient(t, ms)

	if _, err := c.Tools.Invoke(context.Background(), "web.fetch",
		map[string]any{}, plinth.InvokeOpts{Cache: plinth.BoolPtr(false)},
	); err != nil {
		t.Fatalf("Invoke: %v", err)
	}
	var body map[string]any
	ms.LastRequest(t).jsonBody(t, &body)
	if body["cache"] != false {
		t.Errorf("body.cache = %v, want false", body["cache"])
	}
}

// TestToolsRegisterAndList covers the /v1/tools/register +
// /v1/tools listing endpoints.
func TestToolsRegisterAndList(t *testing.T) {
	ms := NewMockServer(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("POST", "/v1/tools/register", 201, map[string]any{
		"tool_id":       "x.tool",
		"name":          "X Tool",
		"description":   "test",
		"transport":     "http",
		"endpoint":      "http://localhost:9000",
		"input_schema":  map[string]any{},
		"output_schema": map[string]any{},
		"created_at":    now,
		"updated_at":    now,
	})
	ms.JSON("GET", "/v1/tools", 200, map[string]any{
		"tools": []any{
			map[string]any{
				"tool_id":       "x.tool",
				"name":          "X Tool",
				"description":   "test",
				"transport":     "http",
				"endpoint":      "http://localhost:9000",
				"input_schema":  map[string]any{},
				"output_schema": map[string]any{},
				"created_at":    now,
				"updated_at":    now,
			},
		},
	})

	c := newTestClient(t, ms)
	tool, err := c.Tools.Register(context.Background(), plinth.ToolRegistration{
		ToolID:       "x.tool",
		Name:         "X Tool",
		Description:  "test",
		Transport:    plinth.ToolTransportHTTP,
		Endpoint:     "http://localhost:9000",
		InputSchema:  map[string]any{},
		OutputSchema: map[string]any{},
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if tool.ToolID != "x.tool" {
		t.Errorf("tool.ToolID = %q, want x.tool", tool.ToolID)
	}

	all, err := c.Tools.List(context.Background())
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(all) != 1 || all[0].ToolID != "x.tool" {
		t.Errorf("List = %v, want one tool x.tool", all)
	}
}

// TestToolsAuditQueryEncodesParams verifies the audit query parameters
// are forwarded as ?workspace_id=&tool_id=&since=&limit=.
func TestToolsAuditQueryEncodesParams(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("GET", "/v1/audit", 200, map[string]any{"events": []any{}})
	c := newTestClient(t, ms)

	if _, err := c.Tools.Audit(context.Background(), plinth.AuditQuery{
		WorkspaceID: "ws_x",
		ToolID:      "web.fetch",
		Since:       "1h",
		Limit:       50,
	}); err != nil {
		t.Fatalf("Audit: %v", err)
	}
	got := ms.LastRequest(t)
	for _, k := range []string{"workspace_id=ws_x", "tool_id=web.fetch", "since=1h", "limit=50"} {
		if !strings.Contains(got.Query, k) {
			t.Errorf("query missing %q (got %q)", k, got.Query)
		}
	}
}

// TestToolsRateLimitedSurfacesRetryAfter verifies that 429 responses
// populate PlinthError.RetryAfter and LimitType.
func TestToolsRateLimitedSurfacesRetryAfter(t *testing.T) {
	ms := NewMockServer(t)
	ms.ErrorWithDetails("POST", "/v1/invoke", 429,
		"RATE_LIMITED", "too fast",
		map[string]any{
			"limit_type":          "rpm",
			"retry_after_seconds": 12.0,
			"current":             60,
			"limit":               60,
		},
	)
	c := newTestClient(t, ms)

	_, err := c.Tools.Invoke(context.Background(), "web.fetch",
		map[string]any{}, plinth.InvokeOpts{},
	)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !errors.Is(err, plinth.ErrRateLimited) {
		t.Errorf("err = %v, want ErrRateLimited", err)
	}
	var pe *plinth.PlinthError
	if !errors.As(err, &pe) {
		t.Fatalf("errors.As to *PlinthError failed: %v", err)
	}
	if pe.RetryAfter != 12.0 {
		t.Errorf("RetryAfter = %v, want 12.0", pe.RetryAfter)
	}
	if pe.LimitType != "rpm" {
		t.Errorf("LimitType = %q, want rpm", pe.LimitType)
	}
}

// TestToolsDryRun covers the /v1/invoke/dry-run endpoint.
func TestToolsDryRun(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("POST", "/v1/invoke/dry-run", 200, map[string]any{
		"tool_id":               "web.fetch",
		"arguments":             map[string]any{"url": "mock://"},
		"would_invoke":          true,
		"estimated_cost_usd":    0.001,
		"estimated_duration_ms": 50,
	})
	c := newTestClient(t, ms)
	resp, err := c.Tools.DryRun(context.Background(), "web.fetch",
		map[string]any{"url": "mock://"}, plinth.InvokeOpts{},
	)
	if err != nil {
		t.Fatalf("DryRun: %v", err)
	}
	if !resp.WouldInvoke {
		t.Errorf("WouldInvoke = false, want true")
	}
	if resp.EstimatedDurationMs != 50 {
		t.Errorf("EstimatedDurationMs = %d, want 50", resp.EstimatedDurationMs)
	}
}
